pkg/scheduler/scheduler.go

```diff
@@ -169,7 +169,8 @@ func (s *Scheduler) schedule(ctx context.Context) {
 				log.Error(err, "Failed to preempt workloads")
 			}
 			if preempted != 0 {
-				e.inadmissibleMsg += fmt.Sprintf(". Preempted %d workload(s)", preempted)
+				e.inadmissibleMsg += fmt.Sprintf(". Pending the preemption of %d workload(s)", preempted)
+				e.requeueReason = queue.RequeueReasonPendingPreemption
```

Comment:
why do we need this reason?

We shouldn't requeue immediately if there is preemption, because actually we need to wait for the running workloads to terminate.

Reply:
That is ok. The users chose BestEffortFIFO, so they should be ok with that. If they want stronger guarantees, they can use StrictFIFO.

Reply:
preempted workloads could take an arbitrary amount of time to fully terminate.

If we put the preemptor back into the head of the queue, this means that we won't admit any other workload until preemptions finish. This is not the behavior that people want when they use BestEffortFIFO. This mode tries to keep throughput up.

The good thing is that when a workload actually terminates, the preemptor workload should get back into the head of the queue.

So I don't think we should specialize preemption when requeueing.

Reply:
I did tried this approach, but the everything gets complicated when multiple workloads should be preempted,  say we have :
```
queue (of 3 cpu)
wl1(low prio, 1 cpu, admitted)
wl2(low prio, 1 cpu, admitted)

create wl3(high prio, 3cpu) 

scheduler step1 (head=wl3):
- wl1 -> evicted
- wl2 -> evicted
- wl3 -> inadmissible

wl1 finish the eviction:
- wl1 -> pending
- wl2 -> evicted
- wl3 -> pending (inadmissible is removed)

scheduler step2 (head=wl3):
- wl1  -> pending (no change)
- wl2 -> evicted
- wl3 -> inadmissible (still needs the cpu used by wl2)

scheduler step3 (head=wl1):
- wl1  -> admitted
- wl2 -> evicted
- wl3 -> inadmissible 

```

sure after this , wl3 will request the preemption of wl1 again, but unless wl1 and wl2 are finishing at the same time, we will be locked in this loop.

Reply:
so we are admitting the recently evicted wl too fast.

This reminds me how we deal with this in kube-scheduler: we would mark wl3 as "nominated", meaning that it's likely to be admitted soon. Then, wl1 wouldn't be able to fit. "nominated" is somewhat similar to "assumed", but with the exception that if the scheduler later takes the decision to put the workload somewhere else, the previous assumption is forgotten.

Could you explore this option? It still gives the chance for other workloads to be admitted, but without taking the space reserved for the preemptor.

Now, the problem you describe can arise even before your change, so we can probably tackle it in a separate PR.

For the purpose of this PR, maybe we can remove RequeueReasonPendingPreemption and change the integration test to only have to preempt 1 workload. We will change it back to 2 once we have implemented the "nominated" logic.

Reply:
> The reservation could work but it will be tricky to implement, since is very likely for us to want to be able to cancel the reservation of lower priority workloads ....

Right, but we would be able to admit another lower priority workload that could fit in a different flavor (assuming the preemptor doesn't fit there).
But sure, this is acceptable.